rbd/diff_iterate_by_id.go

```diff
@@ -0,0 +1,136 @@
+#include <rbd/librbd.h>
+
+#ifdef LIBRBD_SUPPORTS_DIFF_ITERATE3
+#define _RBD_DIFF_ITERATE_FLAG_INCLUDE_PARENT RBD_DIFF_ITERATE_FLAG_INCLUDE_PARENT
+#define _RBD_DIFF_ITERATE_FLAG_WHOLE_OBJECT RBD_DIFF_ITERATE_FLAG_WHOLE_OBJECT
+#define _LIBRBD_SUPPORTS_DIFF_ITERATE3 1
+#else
+#define _RBD_DIFF_ITERATE_FLAG_INCLUDE_PARENT 0
+#define _RBD_DIFF_ITERATE_FLAG_WHOLE_OBJECT 0
+#define _LIBRBD_SUPPORTS_DIFF_ITERATE3 0
+#endif
+
+extern int diffIterateByIDCallback(uint64_t, size_t, int, uintptr_t);
+
+// rbd_diff_iterate3_fn matches the rbd_diff_iterate3 function signature.
+typedef int(*rbd_diff_iterate3_fn)(rbd_image_t image, uint64_t from_snap_id,
+	uint64_t ofs, uint64_t len, uint32_t flags,
+	int (*cb)(uint64_t, size_t, int, void *), void *arg);
+
+// rbd_diff_iterate3_dlsym take *fn as rbd_diff_iterate3_fn and calls the dynamically loaded
+// rbd_diff_iterate3 function passed as 1st argument.
+static inline int rbd_diff_iterate3_dlsym(void *fn, rbd_image_t image,
+	uint64_t from_snap_id, uint64_t ofs, uint64_t len, uint32_t flags, uintptr_t arg) {
+	// cast function pointer fn to rbd_diff_iterate3 and call the function
+	return ((rbd_diff_iterate3_fn) fn)(image, from_snap_id, ofs, len, flags, (void*)diffIterateByIDCallback, (void*)arg);
+}
+*/
+import "C"
+
+import (
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/callbacks"
+	"github.com/ceph/go-ceph/internal/dlsym"
+)
+
+var (
+	diffIterateByIDCallbacks = callbacks.New()
+	diffIterateByIDOnce      sync.Once
+	diffIterateById          unsafe.Pointer
+	diffIterateByIdErr       error
+)
+
+// DiffIterateByIDConfig is used to define the parameters of a DiffIterateByID call.
+// Callback, Offset, and Length should always be specified when passed to
+// DiffIterateByID. The other values are optional.
+type DiffIterateByIDConfig struct {
+	FromSnapID    uint64
+	Offset        uint64
+	Length        uint64
+	IncludeParent DiffIncludeParent
+	WholeObject   DiffWholeObject
+	Callback      DiffIterateCallback
+	Data          interface{}
+}
+
+// DiffIterateByID calls a callback on changed extents of an image.
+//
+// Calling DiffIterateByID will cause the callback specified in the
+// DiffIterateByIDConfig to be called as many times as there are changed
+// regions in the image (controlled by the parameters as passed to librbd).
+//
+// See the documentation of DiffIterateCallback for a description of the
+// arguments to the callback and the return behavior.
+//
+// Implements:
+//
+//	int rbd_diff_iterate3(rbd_image_t image,
+//	                      uint64_t from_snap_id,
+//	                      uint64_t ofs, uint64_t len,
+//	                      uint32_t flags,
+//	                      int (*cb)(uint64_t, size_t, int, void *),
+//	                      void *arg);
+func (image *Image) DiffIterateByID(config DiffIterateByIDConfig) error {
+	if err := image.validate(imageIsOpen); err != nil {
+		return err
+	}
+	if config.Callback == nil {
+		return getError(C.EINVAL)
```

Comment:
```suggestion
		return getError(-C.EINVAL)
```

Reply:
There's no `-` at all lot of places

https://github.com/search?q=repo%3Aceph%2Fgo-ceph%20EINVAL&type=code

This function is almost the same as 
https://github.com/ceph/go-ceph/blob/d284d221f7c09a5968301b9f3cb85cf395d93030/rbd/diff_iterate.go#L98-L100

Let me know if it should have a `-` or not.

Reply:
`getError()` defined in _rbd/errors.go_ has a special case for `ENOENT` with a `-`. It is better to align accordingly with Ceph's negative error code approach.